src/index.js

```diff
@@ -349,131 +278,236 @@ export default class RestClient {
+  async _fetch(method, endpoint, fetchOptions, context, options, metadata = {}) {
     let res, errors;
 
     try {
       logger.info(`${method}: ${this._baseURL}${endpoint}`);
-      let _content;
-      if (content && !isString(content)) _content = JSON.stringify(content);
-
-      res = await fetch(`${this._baseURL}${endpoint}`, {
-        body: _content, headers: new Headers(options.headers), method,
-      });
+      res = await fetch(`${this._baseURL}${endpoint}`, fetchOptions);
     } catch (err) {
       errors = err;
-      logger.error(err);
+      logger.error(errors);
+    }
+
+    const { requests } = this._getTracker(context.path);
+    const fetchID = context.fetchID;
+    if (!requests[fetchID]) requests[fetchID] = [];
+
+    if (errors) {
+      errors = isArray(errors) ? errors : [errors];
+      requests[fetchID].push({ endpoint, errors });
+      return { endpoint, errors };
+    }
+
+    const { headers, ok, status, statusText } = res;
+    const responseGroup = getResponseGroup(status);
+    let _metadata = { ...metadata, endpoint, method, ok, responseGroup, status, statusText };
+    if (!_metadata.errors) _metadata.errors = [];
+    if (!_metadata.resource) _metadata.resource = [];
+
+    if (responseGroup === 'redirection' && headers.get('location')) {
+      if (!_metadata.redirects || _metadata.redirects < 5) {
+        return this._handleRedirect(method, fetchOptions, context, options, headers, _metadata);
+      }
+
+      _metadata.errors.push('The request exceeded the maximum number of redirects.');
+    }
+
+    if (responseGroup === 'serverError' && (!_metadata.retries || _metadata.retries < 3)) {
+      return this._handleRetry(
+        method, endpoint, fetchOptions, context, options, headers, _metadata,
+      );
     }
 
-    if (errors) return { errors, status: res.status };
-    let body = await res[options.streamReader]();
-    body = options.bodyParser(body, context);
-    if (body.errors) return { errors: body.errors, status: res.status };
-    return { data: body.data, headers: res.headers, status: res.status };
+    if (!headers.get('content-type')) {
+      requests[fetchID].push(_metadata);
+      return { headers, ..._metadata };
+    }
+
+    const parsed = options.bodyParser(await res[options.streamReader](), context);
+    _metadata = { ..._metadata };
+
+    if (parsed.errors) {
+      errors = isArray(parsed.errors) ? parsed.errors : [parsed.errors];
+      _metadata.errors = [..._metadata.errors, ...errors];
+    }
+
+    requests[fetchID].push(_metadata);
+    return { data: parsed.data, headers, ..._metadata };
   }
 
   /**
    *
    * @private
    * @param {string} endpoint
-   * @param {string|Array<string>} values
+   * @param {Array<string>} values
    * @param {Object} context
    * @param {Object} options
    * @param {string} resourceKey
    * @return {Promise}
    */
   async _get(endpoint, values, context, options, resourceKey) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 18 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS3776)